### PR TITLE
RSDK-7900 Add resource lifecycle states

### PR DESCRIPTION
**Pull request by maximpertsov**
Replace "state" boolean flags in resource graph nodes with finite states. This change is a no-op that paves the way for exposing more granular information about resource startup and reconfiguration.

### CODE REVIEW

File: resource/graph_node_test.go [hunk 1]
```diff
@@ -183,6 +239,9 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
 	_, err = node.Resource()
 	test.That(t, err, test.ShouldNotBeNil)
 	test.That(t, err.Error(), test.ShouldContainSubstring, "not initialized")
+	// TODO(RSDK-7928): we might want to make this transition a node to an "unconfigured"
+	// or "removing" state.
+	rs = verifySameState(t, node, resource.NodeStateReady, rs)
```
Comment (maximpertsov) [inline resource/graph_node_test.go:244]: making `node.Close` transition to one of the states in the comments results in changes in reconfiguration results, so I'd like to punt the decision to do this for another PR and keep this one as a strict no-op

File: resource/graph_node_test.go [hunk 2]
```diff
@@ -208,6 +268,9 @@ func lifecycleTest(t *testing.T, node *resource.GraphNode, initialDeps []string)
 	_, err = node.Resource()
 	test.That(t, err, test.ShouldNotBeNil)
 	test.That(t, err.Error(), test.ShouldContainSubstring, "not initialized")
+	// TODO(RSDK-7928): we might want to make this transition a node to an "unconfigured"
+	// or "removing" state.
+	verifySameState(t, node, resource.NodeStateReady, rs)
```
Comment (maximpertsov) [inline resource/graph_node_test.go:273]: [see comment](https://github.com/viamrobotics/rdk/pull/4075/files#r1639971960)

File: resource/graph_node.go [hunk 3]
```diff
@@ -356,24 +397,35 @@ func (w *GraphNode) replace(other *GraphNode) error {
...
 	return nil
 }
+
+func (w *GraphNode) transitionTo(state NodeState) {
+	if w.state == state {
+		return
+	}
+	w.state = state
+	w.transitionedAt = time.Now()
+}
```
Comment (maximpertsov) [inline resource/graph_node.go:431]: no concrete follow-up, but two notes/thoughts:
* we could consider expanding this helper to validate if a transition is legal
* it might be interesting to make self-transitions error to expose scenarios where need a more granular state

File: resource/graph_node.go [hunk 4]
```diff
@@ -11,6 +11,24 @@ import (
 	"go.viam.com/rdk/logging"
 )
 
+// NodeState captures the configuration lifecycle state of a resource node.
+type NodeState uint8
+
+const (
+	// NodeStateUnconfigured denotes a newly created resource.
+	NodeStateUnconfigured NodeState = iota
```
Comment (cheukt) [inline resource/graph_node.go:19]: prefer if we just kept it 1-1 to proto - so Initialize, Configure, Ready, Remove, Unhealthy.

File: resource/graph_node.go [hunk 5]
```diff
@@ -111,6 +133,24 @@ func (w *GraphNode) Resource() (Resource, error) {
 	return w.current, nil
 }
 
+// WithState contains a resource along with lifecycle state information.
+type WithState struct {
```
Comment (cheukt) [inline resource/graph_node.go:137]: what will this be used for?

File: resource/graph_node.go (hunk 3 above)
Comment (cheukt) [reply resource/graph_node.go:431]: agreed, I think we should eventually use this helper to validate if a transition is legal

File: resource/graph_node.go (hunk 4 above)
Comment (maximpertsov) [reply resource/graph_node.go:19]: Oh, I was planning to make these the proto states as well. I think `Unconfigured`, `Configuring`, `Removing` are more "stateful" descriptors. The original ones sound more like commands.

File: resource/graph_node.go (hunk 5 above)
Comment (maximpertsov) [reply resource/graph_node.go:137]: Right now just for tests, but will eventually be exported in the robot API

File: resource/graph_node.go (hunk 4 above)
Comment (maximpertsov) [reply resource/graph_node.go:19]: Updated related tickets with a comment: https://viam.atlassian.net/browse/RSDK-7900?focusedCommentId=31236, https://viam.atlassian.net/browse/RSDK-7899

Let me know if you disagree

File: resource/graph_node.go [hunk 6]
```diff
@@ -11,6 +11,24 @@ import (
...
+
+const (
+	// NodeStateUnconfigured denotes a newly created resource.
+	NodeStateUnconfigured NodeState = iota
+
+	// NodeStateConfiguring denotes a resource is being configured.
+	NodeStateConfiguring
+
+	// NodeStateReady denotes a resource that has been initialized and is not being
+	// configured or removed. A resource in this state may be unhealthy.
```
Comment (benjirewis) [inline resource/graph_node.go:25]: > A resource in this state may be unhealthy.

Why is there no `NodeStateUnhealthy`?

File: resource/graph_node.go [hunk 7]
```diff
@@ -32,14 +50,18 @@ type GraphNode struct {
...
 	unresolvedDependencies    []string
 	needsDependencyResolution bool
 
 	logger logging.Logger
+
+	// state stores the current lifecycle state for a resource node
+	state NodeState
+	// transitionedAt stores the timestamp of when resource entered it's current
+	// lifecycle state.
+	transitionedAt time.Time
```
Comment (benjirewis) [inline resource/graph_node.go:64]: Maybe a `*time.Time` as for `lastReconfigured`? The zero-value of a `time.Time` is something like Jan 1. 1970.

File: resource/graph_node.go [hunk 8]
```diff
@@ -32,14 +50,18 @@ type GraphNode struct {
 	current                   Resource
 	currentModel              Model
 	config                    Config
-	needsReconfigure          bool
 	lastReconfigured          *time.Time
```
Comment (benjirewis) [inline resource/graph_node.go:53]: Would you mind adding a comment above `lastReconfigured` stating its purpose (particularly to distinguish it from `transitionedAt`)? Something like:
```
lastReconfigured stores the timestamp at which the resource was constructed or last reconfigured. 
transitionedAt is generally a more useful timestamp, but `lastReconfigured` is still returned as part
of resource status.
```
`transitionedAt` and the `MachineStatus` endpoint will be a lot more informative than `lastReconfigured` and the `Status` endpoint.

File: resource/graph_node.go [hunk 9]
```diff
@@ -111,6 +133,24 @@ func (w *GraphNode) Resource() (Resource, error) {
...
 
+// WithState contains a resource along with lifecycle state information.
+type WithState struct {
+	Resource  Resource
+	State     NodeState
+	UpdatedAt time.Time
+}
+
+// ResourceWithState returns the underlying resource along with its current state.
+func (w *GraphNode) ResourceWithState() WithState {
```
Comment (benjirewis) [inline resource/graph_node.go:144]: This is odd API to me. Why not have a `State()` method and a `TransitionedAt()` method? I suppose I'm confused why we're returning what is effectively a tuple instead of having separate methods, and why `transitionedAt` became `UpdatedAt`.

File: resource/graph_node.go (hunk 3 above)
Comment (benjirewis) [reply resource/graph_node.go:431]: Really like both those thoughts. Why not add both of those now?

File: resource/graph_node_test.go [hunk 10]
```diff
@@ -74,10 +80,44 @@ func TestConfiguredLifecycle(t *testing.T) {
...
+
 	lifecycleTest(t, node, []string(nil))
 }
 
+func verifyStateTransition(
+	t *testing.T,
+	node *resource.GraphNode,
+	expectedState resource.NodeState,
+	lastState resource.WithState,
+) resource.WithState {
```
Comment (benjirewis) [inline resource/graph_node_test.go:94]: [nit] It looks like you don't used the returned value here? Should we just return nothing from these helpers?

**Review by benjirewis — COMMENTED** [review]
Very cool 🧑‍🔧 ; a few initial qs/comments.

File: resource/graph_node.go (hunk 3 above)
Comment (cheukt) [reply resource/graph_node.go:431]: if we're worried that some transitions are illegal, we should log a warning for now and try to root it out

Comment (benjirewis) [reply resource/graph_node.go:431]: Logging for now SGTM.

File: resource/graph_node.go (hunk 5 above)
Comment (cheukt) [reply resource/graph_node.go:137]: agree with @benjirewis  [here](https://github.com/viamrobotics/rdk/pull/4075/files#r1643234587), feels like we could just expose GraphNode without going through the extra steps

File: resource/graph_node.go [hunk 11]
```diff
@@ -11,6 +11,24 @@ import (
...
+	// NodeStateReady denotes a resource that has been initialized and is not being
+	// configured or removed. A resource in this state may be unhealthy.
+	NodeStateReady
+
+	// NodeStateRemoving denotes a resource is being removed from the resource graph.
+	NodeStateRemoving
+)
+
 // A GraphNode contains the current state of a resource.
 // It starts out as either uninitialized, unconfigured, or configured.
```
Comment (cheukt) [inline resource/graph_node.go:33]: should we also add `uninitialized` or is that an impossible state?

File: resource/graph_node.go (hunk 4 above)
Comment (cheukt) [reply resource/graph_node.go:19]: i think the name changes are fine, just surprised they changed lol

Comment (maximpertsov) [reply resource/graph_node.go:19]: yeah, naming stuff is hard :name_badge:

File: resource/graph_node.go (hunk 11 above)
Comment (maximpertsov) [reply resource/graph_node.go:33]: it's not clear to me how this state would be different from `unconfigured` (i.e. the default state) as currently proposed - right now i'm using `unconfigured` to capture everything before `configuring`

File: resource/graph_node.go (hunk 9 above)
Comment (maximpertsov) [reply resource/graph_node.go:144]: > tuple instead of having separate methods

I bundled them together to avoid consistency issues related to accessing one field then the other when an update happens in between. my thought is that if you are getting a state you generally want update timestamp as well. In retrospect, maybe the field should just hold a struct then.

> why transitionedAt became UpdatedAt

I wanted to name this field `updated_at` initially but there is another field with the same name that is used for testing.

File: resource/graph_node.go (hunk 6 above)
Comment (maximpertsov) [reply resource/graph_node.go:25]: it's a [TODO](https://github.com/viamrobotics/rdk/pull/4075/files#diff-d6f0241bd713d1ea97a4752dbb520099399b7b2edd9e94d539d0674e83cefedfR266) - I want to avoid having references to it until we actually plan to implement.

File: resource/graph_node.go (hunk 3 above)
Comment (maximpertsov) [reply resource/graph_node.go:431]: great idea, will add warning logs in these cases

File: resource/graph_node_test.go (hunk 10 above)
Comment (maximpertsov) [reply resource/graph_node_test.go:94]: [i am using them](https://github.com/viamrobotics/rdk/pull/4075/files#diff-4bd90405fd75a317db2dbde82c00b5dbec5dcef93fb63725af573afbb4603d96R130), but maybe it's confusing... my aim here is that the given `lastState` is compared against the current state, and then the current state is returned to be asserted against at the next point in the test where we expect to see a transition

File: resource/graph_node.go (hunk 11 above)
Comment (cheukt) [reply resource/graph_node.go:33]: can you update the docstring then?

File: resource/graph_node.go (hunk 5 above)
Comment (maximpertsov) [reply resource/graph_node.go:137]: true, i guess `GraphNode` is already supposed to be "resource with state"

File: resource/graph_node.go (hunk 11 above)
Comment (maximpertsov) [reply resource/graph_node.go]: somewhat related, it might be worth having an "unknown" state as the default/sentinel value. this value would always be treated as an error

File: resource/graph_node.go [hunk 12]
```diff
@@ -359,24 +407,71 @@ func (w *GraphNode) replace(other *GraphNode) error {
...
+		switch state {
+		case NodeStateConfiguring, NodeStateRemoving:
+			return true
+		}
+	case NodeStateRemoving:
+	}
+	return false
+}
+
+func (w *GraphNode) transitionTo(state NodeState) {
```
Comment (cheukt) [inline resource/graph_node.go:461]: leave a comment that this should be called within a lock?

**Review by cheukt — APPROVED** [review]
LGTM

File: resource/graph_node.go (hunk 3 above)
Comment (maximpertsov) [reply resource/graph_node.go]: I didn't dive too deep into this, but here are the possibly illegal transitions that happen in our test suite. i'm not counting self-transitions since it's clear if that should be okay yet.

|Transition|Possible reason|
|--|--|
|Configuring -> Removing |Seems like resources can "abort" configuration if they are marked for removal. maybe this is fine. |
| Removing -> Configuring | Allowed due to [this logic](https://github.com/viamrobotics/rdk/blob/053632958fea4fbff817bd2ec3a420265fc50ac9/resource/graph_node.go#L259-L276), which forces nodes marked for removal to be configured instead. |
| Removing -> Ready | This one is weird - unclear why how this can happen.|

File: resource/graph_node.go [hunk 13]
```diff
@@ -11,8 +11,30 @@ import (
 	"go.viam.com/rdk/logging"
 )
 
+//go:generate stringer -type NodeState -trimprefix NodeState
```
Comment (benjirewis) [inline resource/graph_node.go:14]: Cool; I've never seen `go:generate` used before. Have you used it for enums before/why is it better than just writing a `String` method?

**Review by benjirewis — APPROVED** [review]
This is great! Thanks 🧑‍🔧

File: resource/graph_node.go (hunk 13 above)
Comment (maximpertsov) [reply resource/graph_node.go:14]: Sorry missed this comment.

This stringer package basically writes the String method for you and ensures all enumerated cases are covered. It supports some basic transformations like trimming parts of the variable name or hard-coding a one-off string for a particular entry.

Comment (benjirewis) [reply resource/graph_node.go:14]: Makes sense; I suppose valuable for enums if we add more variants in the future.
